pkg/scaffold/v2/main.go

```diff
@@ -45,6 +47,67 @@ func (m *Main) GetInput() (input.Input, error) {
 	return m.Input, nil
 }
 
+// Update updates main.go with code fragments required to wire a new
+// resource/controller.
+func (m *Main) Update(opts *MainUpdateOptions) error {
+	path := "main.go"
+
+	resPkg, _ := getResourceInfo(opts.Resource, input.Input{
+		Domain: opts.Project.Domain,
+		Repo:   opts.Project.Repo,
+	})
+
+	// generate all the code fragments
+	apiImportCodeFragment := fmt.Sprintf(`%s%s "%s/%s"
+`, opts.Resource.Group, opts.Resource.Version, resPkg, opts.Resource.Version)
+	ctrlImportCodeFragment := fmt.Sprintf(`"%s/controllers"
+`, opts.Project.Repo)
+	addschemeCodeFragment := fmt.Sprintf(`%s%s.AddToScheme(scheme)
+`, opts.Resource.Group, opts.Resource.Version)
+	reconcilerSetupCodeFragment := fmt.Sprintf(`err = (&controllers.%sReconciler{
+	 	Client: mgr.GetClient(),
+        Log: ctrl.Log.WithName("controllers").WithName("%s"),
+	 }).SetupWithManager(mgr)
+	 if err != nil {
+	 	setupLog.Error(err, "unable to create controller", "controller", "%s")
+	 	os.Exit(1)
+	 }
+`, opts.Resource.Kind, opts.Resource.Kind, opts.Resource.Kind)
+
+	if opts.WireResource {
+		err := internal.InsertStringsInFile(path,
+			apiPkgImportScaffoldMarker, apiImportCodeFragment,
+			apiSchemeScaffoldMarker, addschemeCodeFragment)
+		if err != nil {
+			return err
+		}
+	}
+
+	if opts.WireController {
+		return internal.InsertStringsInFile(path,
+			apiPkgImportScaffoldMarker, apiImportCodeFragment,
+			apiPkgImportScaffoldMarker, ctrlImportCodeFragment,
+			apiSchemeScaffoldMarker, addschemeCodeFragment,
+			reconcilerSetupScaffoldMarker, reconcilerSetupCodeFragment)
+	}
+
+	return nil
+}
+
+// MainUpdateOptions contains info required for wiring an API/Controller in
+// main.go.
+type MainUpdateOptions struct {
+	// Project contains info about the project
+	Project *input.ProjectFile
+
+	// Resource is the resource being added
+	Resource *resource.Resource
```

Comment:
`resource.Resource` comes from `"sigs.k8s.io/kubebuilder/pkg/scaffold/v1/resource"`.
IMO a `v2` package should not depend on a `v1` package. IOW, `resource` should be moved to a shared package.

Reply:
> IMO a v2 package should not depend on a v1 package.

Thats debatable :). If you don't feel strongly about it, I would keep this refactor out of scope for this PR.

Reply:
IMHO when I change the scaffolding under v1 package, I wouldn't expect v2 scaffolding being affected. Otherwise, it makes it hard to reasoning.

I did a quick check, it seems we have quite some places in v2 that are importing v1 packages. It makes me sad. 
The change here is not the first one doing it. I'm OK to leave it in a separate PR. :)

Reply:
I took a look at the use of v1.Resource in v2 scaffolding and it turns out data to represent a resource is entirely same when it comes to v1/v2 scaffolding, so if we try to completely isolate v1/v2, we will end up duplicating the structure. Now for sharing the same data structure, one could move v1.Resource to a common pkg, but that has the same affect of importing v1 pkg in v2 because one has to reason about the usage of that shared structure in across v1/v2 scaffolding.
WDYT ?

Reply:
Reorganize the code into `common`, `v1` and `v2` can make it easier to maintain the scaffolding the code in long term. But if adding change something in `common`, contributors still need to reason the usage of share structure across v1 and v2 scaffolding.

We can keep the scope of this PR as is, and have these thoughts in mind when we reorganizing the code in the followup PR.

Reply:
Ack.